apollo-router/src/services/connector_service/snapshot.rs

```diff
@@ -0,0 +1,122 @@
+//! Snapshot responses from REST APIs for Connectors.
+//!
+//! Snapshots store the response from a REST API (headers and body) to the local file system.
+//! The response can then be reused for subsequent requests to the same URL. The files are JSON,
+//! and can be manually edited, for example to redact or remove sensitive fields or headers, or
+//! to change values for testing.
+//!
+//! Snapshots are useful for:
+//! * Development and debugging without repeatedly hitting the backend REST service
+//! * Development and debugging where short-lived auth tokens are required by the REST API
+//! * Creating tests that capture output from a real REST API to replay as a mock
+//!
+//! A few things to be aware of:
+//! * Saving REST responses to the file system could potentially result in saving sensitive
+//!   information in plain text on the file system. This feature is for development purposes
+//!   and should not be used with sensitive production data (though snapshot content can be
+//!   manually redacted as mentioned above). There is a warning emitted on router startup
+//!   when this feature is turned on.
+//! * Snapshots for requests that would modify data (such as POST requests) will obviously
+//!   not have any side effects when replayed. Snapshots are not useful if you rely on those
+//!   side effects.
+//! * The unique key for storing and loading snapshots is the full URL. This means snapshots
+//!   will not work well with APIs running on ephemeral ports, or any other cases where the
+//!   URL might be different for the same request.
+//! * Since headers are not included in the snapshot key, requests that would return
+//!   different data depending on the value of headers will use the same snapshot.
+//! * Since no network request is made, running with snapshots will be unrealistically fast,
+//!   and should not be used with performance or load testing.
+
+use std::borrow::Cow;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use http::HeaderMap;
+use http::HeaderName;
+use http::HeaderValue;
+use indexmap::IndexMap;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_json_bytes::Value;
+use tracing::warn;
+
+use crate::services::router::body::RouterBody;
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct Snapshot<'a> {
+    url: Cow<'a, str>,
+    headers: IndexMap<String, String>,
+    body: Cow<'a, Value>,
+}
+
+impl<'a> Snapshot<'a> {
+    pub(crate) fn new(url: &'a str, body: &'a Value, headers: &'a HeaderMap<HeaderValue>) -> Self {
+        let headers = headers
```

Comment:
should this be a `Vec<(String, String)>` to support repeated header keys?

Reply:
Good call. The `HeaderMap` doesn't support serialization so I had to switch map types, but I overlooked that `HeaderMap` is a multi-map.